### ISSUE TITLE
crt-static is only translated to `-static` under gcc, not clang

### ISSUE THREAD

**Issue by mqudsi** [rust-lang/cc-rs#1074]
In `add_default_flags()`, the presence of feature `crt-static` causes `-static` to be supplied to the compiler [iff it's gcc](https://github.com/rust-lang/cc-rs/blob/3ba23569a623074748a3030f382afd22483555df/src/lib.rs#L2173-L2175); clang doesn't get the same treatment.

I know that `cc` itself doesn't compile without `-c` so the presence of `-static` makes no difference there, but if a downstream library or crate uses `cc` to manually get a compiler invocation and runs that without `-c` to generate an executable itself in `build.rs` (and plenty do), it matters.

**Comment by NobodyXu** [comment]
Hmmm, could we remove that `-static` flag?

I wonder why is that added in the first place.

**Comment by mqudsi** [comment]
I think removing it makes the most sense. `crt-static` is not `static` and shouldn't come with the assumption that just because we're statically linking the runtime it means that all libraries should also be statically linked. 

I mean, that's one way of interpreting it but clearly rust *doesn't* interpret it that way, so we probably shouldn't either. 

Not sure about breakage, though.

**Comment by NobodyXu** [comment]
> Not sure about breakage, though. 

I think it should be fine.

Don't think there are many crates using cc to generate executables

**Comment by NobodyXu** [comment]
Looking at it, I found that we also have a function `cc::Build::static_flag` to add `-static`.

Honestly I have no idea why it's there in the first place, cc @thomcc Do you know why it is there (and if we could just remove https://github.com/rust-lang/cc-rs/blob/3ba23569a623074748a3030f382afd22483555df/src/lib.rs#L2173-L2175 )?

**Comment by mqudsi** [comment]
I don't think you can just remove static_flag without breaking a decent amount of code. [It's used a decent amount](https://github.com/search?q=static_flag+path%3A**%2Fbuild.rs+language%3ARust&ref=advsearch&type=code) though as far as I can see it seems most usage is under the very incorrect assumption that it determines whether the output is a .a or .so/.dylib. 

We should be able to safely remove the flag itself from default_flags() though and mark the static_flag() routine as deprecated?

**Comment by NobodyXu** [comment]
> and separately mark the static_flag() routine as deprecated

I agree, it should be marked as deprecated, while keep the existing implementation.

I will then open a PR to remove https://github.com/rust-lang/cc-rs/blob/3ba23569a623074748a3030f382afd22483555df/src/lib.rs#L2173-L2175

**Comment by riidefi** [comment]
We should probably specify the clang-equivalent of `-MT` (on clang-cl) when on Windows. It is odd that if you run `SET CXX=clang` before `cargo build` on Windows, the `-Ctarget-feature=+crt-static` RUSTFLAG is ignored.

E.g.
> Tell C++ to use the dynamic version, via one of the following methods (depending on build system):
> - Directly specify '/MD' to cl.exe or clang-cl.exe during the C++ compile/link steps.
> - Directly specify '-Wl,-nodefaultlib:libcmt -D_DLL -lmsvcrt' to clang++.exe

**Comment by polarathene** [comment]
> `crt-static` is not `static` and shouldn't come with the assumption that just because we're statically linking the runtime it means that all libraries should also be statically linked.

Correct me if I'm mistaken, but if you static link the runtime you no longer have the ability to do `dlopen()` calls, so dynamic linking is not available.

---

## More details

It is problematic to try static link a library while dynamic link another library with overlapping deps, that causes a symbol conflict doesn't it? So instead you may continue to dynamic link libc, but optionally static link in some libs to prefer those over dynamic linking to the system/external libs.

As demonstrated in this reproduction below, a x86_64 glibc build environment can do a "static" glibc build, but:
- It cannot dynamically link to zlib when `+crt-static`.
- It can dynamically link zlib when `-crt-static`.
- It can statically link zlib regardless of `+crt-static` / `-crt-static`.

Without `+crt-static`, dynamic linking will be preferred by default but if only a static library is available it will link to that.
- This can be changed if adding a [specific link kind](https://doc.rust-lang.org/rustc/command-line-arguments.html#-l-link-the-generated-crate-to-a-native-library).
- With compilers that `cc` would use, the ability to do similar with dynamic vs static link kinds is less reliable.
  - `zig cc` seems to ignore standard dynamic/static linker args, it's `-lz` does behave the same as the rust equivalent preferring `libz.so` over `libz.a`, but there is no `-static` support (_this is arriving in Zig v15 but not for non-native / custom-glibc targets_), and the alternate syntax `-l:libz.a` is not supported with `-target` usage (_presumably similar to the `-static` support constraints_).

**So to recap:**
- Isn't `+crt-static` a given that no dynamic linking will be supported? Or is this not the case with some other platforms? (_I've only tested linux gnu/musl_)
- Crates may offer `static`/`vendored` features to minimize dynamically linked deps, plus not all crates can support fully static linked binaries.
  - For example, using `cudarc` needs to dynamic link the `libcuda.so` library due to `dlopen()` usage, `cudarc` supports static linking other CUDA libs that offer pre-built static archives, but those can require the [`whole-archive` modifier](https://doc.rust-lang.org/rustc/command-line-arguments.html#linking-modifiers-whole-archive) due to an issue with `libstdc++` symbols IIRC.

---

## Reproduction for `+crt-static` impact on linking

```bash
docker run --rm -it --workdir /example fedora:42
dnf install -yq cargo glibc-static
dnf --installroot /sysroot/amd64 --use-host-config install -yq zlib-static

cargo init
echo 'fn main() { println!("cargo:rustc-link-lib=z"); }' > build.rs
mkdir -p libz/{static,dynamic}
ln -s /sysroot/amd64/usr/lib64/libz.a libz/static/libz.a
ln -s /sysroot/amd64/usr/lib64/libz.so libz/dynamic/libz.so

# Add a basic lib call to ensure lib is actually linked instead of DCE'd:
echo <<HEREDOC > src/main.rs
use core::ffi::c_char;
unsafe extern "C" {
    #[link_name = "zlibVersion"]
    pub fn zlibVersion() -> *const c_char;
}

fn main() {
    println!("Hello, world! {:?}", unsafe { zlibVersion() });
}
HEREDOC
```

Standard build, zlib not available in path to link fails:

```console
$ cargo build

   Compiling example v0.1.0 (/example)
error: linking with `cc` failed: exit status: 1
  |
  = note:  "cc" "-m64" "/tmp/rustciP3Wt7/symbols.o" "<7 object files omitted>" "-Wl,--as-needed" "-Wl,-Bdynamic" "-lz" "-Wl,-Bstatic" "<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/lib/{libstd-*,libpanic_unwind-*,libobject-*,libmemchr-*,libaddr2line-*,libgimli-*,librustc_demangle-*,libstd_detect-*,libhashbrown-*,librustc_std_workspace_alloc-*,libminiz_oxide-*,libadler2-*,libunwind-*,libcfg_if-*,liblibc-*,liballoc-*,librustc_std_workspace_core-*,libcore-*,libcompiler_builtins-*}.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-L" "/tmp/rustciP3Wt7/raw-dylibs" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/example/target/debug/deps/example-5d298dac7be886a3" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-nodefaultlibs"
  = note: some arguments are omitted. use `--verbose` to show all linker arguments
  = note: /usr/sbin/ld: cannot find -lz: No such file or directory
          collect2: error: ld returned 1 exit status
```

Static build via `+crt-static` with `zlib.so` discoverable, **fails** as static lib is queried:

```console
$ RUSTFLAGS='-L libz/dynamic -C target-feature=+crt-static' cargo build

   Compiling example v0.1.0 (/example)
error: linking with `cc` failed: exit status: 1
  |
  = note:  "cc" "-m64" "/tmp/rustcIC4lPk/symbols.o" "<7 object files omitted>" "-Wl,--as-needed" "-Wl,-Bstatic" "-lz" "<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/lib/{libstd-*,libpanic_unwind-*,libobject-*,libmemchr-*,libaddr2line-*,libgimli-*,librustc_demangle-*,libstd_detect-*,libhashbrown-*,librustc_std_workspace_alloc-*,libminiz_oxide-*,libadler2-*,libunwind-*,libcfg_if-*,liblibc-*}.rlib" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-lgcc_eh" "-lgcc" "-lc" "<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/lib/{liballoc-*,librustc_std_workspace_core-*,libcore-*,libcompiler_builtins-*}.rlib" "-L" "/tmp/rustcIC4lPk/raw-dylibs" "-Wl,-Bdynamic" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "libz/dynamic" "-L" "<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/example/target/debug/deps/example-9ab488d81598b379" "-Wl,--gc-sections" "-static-pie" "-Wl,-z,relro,-z,now" "-nodefaultlibs"
  = note: some arguments are omitted. use `--verbose` to show all linker arguments
  = note: /usr/sbin/ld: cannot find -lz: No such file or directory
          /usr/sbin/ld: have you installed the static version of the z library ?
          collect2: error: ld returned 1 exit status
```

Switch to `libz.a` being discoverable, build is successful:

```console
$ RUSTFLAGS='-L libz/static -C target-feature=+crt-static' cargo build

   Compiling example v0.1.0 (/example)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 1.52s

$ ldd target/debug/example
        statically linked
```

Again but switch to `-crt-static`, only `libz.a` is discoverable so that remains statically linked:

```console
$ rm -rf target
$ RUSTFLAGS='-L libz/static -C target-feature=-crt-static' cargo build
   Compiling example v0.1.0 (/example)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 3.65s

$ ldd target/debug/example
        linux-vdso.so.1 (0x00007fffcc373000)
        libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f2e3d7c7000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f2e3d5d5000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f2e3d851000)
```

Switch to dynamic location to find `libz.so` instead, now zlib is linked dynamically:

```console
$ rm -rf target
$ RUSTFLAGS='-L libz/dynamic -C target-feature=-crt-static' cargo build
   Compiling example v0.1.0 (/example)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.81s

$ ldd target/debug/example
        linux-vdso.so.1 (0x00007ffe62afe000)
        libz.so.1 => /lib64/libz.so.1 (0x0000716355f03000)
        libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x0000716355ed7000)
        libc.so.6 => /lib64/libc.so.6 (0x0000716355ce5000)
        /lib64/ld-linux-x86-64.so.2 (0x0000716355f84000)
```

**Comment by NobodyXu** [comment]
> Isn't +crt-static a given that no dynamic linking will be supported? Or is this not the case with some other platforms? (I've only tested linux gnu/musl)

Hmm so on linux gnu/musl it's not necessary to pass `-static` if `+crt-static`
